sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/AvroWrapperCoder.java

```diff
@@ -0,0 +1,115 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A {@code AvroWrapperCoder} is a {@link org.apache.beam.sdk.coders.Coder} for a Java class that
+ * implements {@link org.apache.avro.mapred.AvroWrapper}.
+ *
+ * @param <WrapperT> the type of the wrapper
+ * @param <DatumT> the type of the datum
+ */
+public class AvroWrapperCoder<WrapperT extends AvroWrapper<DatumT>, DatumT>
+    extends StandardCoder<WrapperT> {
+  private static final long serialVersionUID = 0L;
+
+  private final Class<WrapperT> wrapperType;
+  private final AvroCoder<DatumT> datumCoder;
+
+  private AvroWrapperCoder(Class<WrapperT> wrapperType, AvroCoder<DatumT> datumCoder) {
+    this.wrapperType = wrapperType;
+    this.datumCoder = datumCoder;
+  }
+
+  /**
+   * Return a {@code AvroWrapperCoder} instance for the provided element class.
+   * @param <WrapperT> the type of the wrapper
+   * @param <DatumT> the type of the datum
+   */
+  public static <WrapperT extends AvroWrapper<DatumT>, DatumT>
+  AvroWrapperCoder<WrapperT, DatumT>of(Class<WrapperT> wrapperType, AvroCoder<DatumT> datumCoder) {
+    return new AvroWrapperCoder<>(wrapperType, datumCoder);
+  }
+
+  @JsonCreator
+  @SuppressWarnings("unchecked")
+  public static AvroWrapperCoder<?, ?> of(
+          @JsonProperty("wrapperType") String wrapperType,
+          @JsonProperty(PropertyNames.COMPONENT_ENCODINGS) List<Coder<?>> components)
+          throws ClassNotFoundException {
+    Class<?> clazz = Class.forName(wrapperType);
+    if (!AvroWrapper.class.isAssignableFrom(clazz)) {
+      throw new ClassNotFoundException(
+              "Class " + wrapperType + " does not implement AvroWrapper");
+    }
+    Preconditions.checkArgument(components.size() == 1,
+            "Expecting 1 component, got " + components.size());
+    return of((Class<? extends AvroWrapper>) clazz, (AvroCoder<?>) components.get(0));
+  }
+
+  @Override
+  public void encode(WrapperT value, OutputStream outStream, Context context) throws IOException {
+    datumCoder.encode(value.datum(), outStream, context);
+  }
+
+  @Override
+  public WrapperT decode(InputStream inStream, Context context) throws IOException {
+    try {
+      WrapperT wrapper = wrapperType.newInstance();
+      wrapper.datum(datumCoder.decode(inStream, context));
+      return wrapper;
+    } catch (InstantiationException | IllegalAccessException e) {
+      throw new CoderException("unable to deserialize record", e);
+    }
+  }
+
+  @Override
+  public List<? extends Coder<?>> getCoderArguments() {
+    return Collections.singletonList(datumCoder);
+  }
+
+  @Override
+  public CloudObject asCloudObject() {
+    CloudObject result = super.asCloudObject();
+    result.put("wrapperType", wrapperType.getName());
+    return result;
+  }
+
+  @Override
+  public void verifyDeterministic() throws NonDeterministicException {}
```

Comment:
`WrapperT` is never actually serialized so it shouldn't matter but you're right about AvroCoder, so I changed this to delegate to `datumCoder.verifyDeterministic();`